manticore/platforms/evm_world_state.py

```diff
@@ -0,0 +1,445 @@
+        self.warned = False
+        self.map = constraints.new_array(
+            index_bits=256,
+            value_bits=1,
+            name=f"STORAGE_MAP_{address:x}",
+            avoid_collisions=True,
+            default=0,
+        )
+        self.data = constraints.new_array(
+            index_bits=256,
+            value_bits=256,
+            name=f"STORAGE_DATA_{address:x}",
+            avoid_collisions=True,
+            default=0,
+        )
+        self.dirty = False
+
+    def set(self, offset: Union[int, BitVec], value: Union[int, BitVec]):
+        self.map[offset] = 1
+        self.data[offset] = value
+        self.dirty = True
+
+    def dump(self, stream: TextIOBase, state: State):
+        concrete_indexes = set()
+        for sindex in self.map.written:
+            concrete_indexes.add(state.solve_one(sindex, constrain=True))
+
+        for index in concrete_indexes:
+            stream.write(
+                f"storage[{index:x}] = {state.solve_one(self.data[index], constrain=True):x}"
+            )
+
+    @staticmethod
+    def from_dict(constraints: ConstraintSet, address: int, items: Dict[int, int]) -> "Storage":
+        storage = Storage(constraints, address)
+        for key, value in items.items():
+            storage.set(key, value)
+        return storage
+
+
+####################################################################################################
+
+
+class WorldState:
+    @abstractmethod
+    def is_remote(self) -> bool:
+        pass
+
+    @abstractmethod
+    def accounts(self) -> Set[int]:
+        pass
+
+    @abstractmethod
+    def get_nonce(self, address: int) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def get_balance(self, address: int) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def has_storage(self, address: int) -> bool:
+        pass
+
+    @abstractmethod
+    def get_storage(self, address: int) -> Optional[Storage]:
+        pass
+
+    @abstractmethod
+    def get_storage_data(self, address: int, offset: Union[int, BitVec]) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def get_code(self, address: int) -> Union[bytes, Array]:
+        pass
+
+    @abstractmethod
+    def get_blocknumber(self) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def get_timestamp(self) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def get_difficulty(self) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def get_gaslimit(self) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def get_coinbase(self) -> Union[int, BitVec]:
+        pass
+
+
+####################################################################################################
+
+
+class DefaultWorldState(WorldState):
+    def is_remote(self) -> bool:
+        return False
+
+    def accounts(self) -> Set[int]:
+        return set()
+
+    def get_nonce(self, address: int) -> int:
+        return 0
```

Comment:
Maybe this should raise not implemented? 
I think some defaults are different?

Reply:
When would the default be different?

Reply:
There was an EIP that defined different initial nonces for contracts vs. normal accounts but that should be enforced upstream probably

Reply:
It looks like we are doing the right thing:
https://github.com/trailofbits/manticore/blob/88ddfbca9efcaa47f736183bee4fd4c6d3ef9b66/manticore/platforms/evm.py#L2779-L2781
Nonetheless, I added some checks to `test_general.py`:
https://github.com/trailofbits/manticore/pull/1580/commits/88ddfbca9efcaa47f736183bee4fd4c6d3ef9b66